exonum/src/blockchain/transaction.rs

```diff
@@ -565,72 +650,99 @@ macro_rules! transactions {
-                            ::deserialize(&value)
-                            .map_err(|e| D::Error::custom(
-                                format!("Can't deserialize a value: {}", e.description())
-                            ))
-                            .map($transaction_set::$name),
-                    )*
-                    _ => Err(D::Error::custom(format!("invalid message_id: {}", message_id))),
-                }
-            }
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __enum_to_vec_id{
+    ($set:ident $this:expr, $($field_type:ident)*) => {
+        __enum_to_vec_id!(@inner $set $this, (0) (); $($field_type)* );
+    };
+    (@create $set:ident $this:expr, ( $((($name:ident) => ($num:expr)))* ) ) => {
+        match $this {
+            $(
+                &$set::$name(ref tx) => ($num, $crate::messages::BinaryForm::encode(tx).unwrap())
+            ),*
         }
+    };
+    (@inner $set:ident $this:expr, ($num:expr) ($($processed:tt)*);
+        $field_type:ident $($rest:tt)*
+    ) => {
+        __enum_to_vec_id!(
+            @inner $set $this,
+            ($num + 1)
+            ($($processed)* (($field_type) => ($num)));
+            $($rest)*
+        );
+    };
 
-        impl $crate::encoding::serialize::reexport::Serialize for $transaction_set {
-            fn serialize<S>(&self, serializer: S) -> ::std::result::Result<S::Ok, S::Error>
-            where
-                S: $crate::encoding::serialize::reexport::Serializer,
-            {
-                use $crate::encoding::serialize::reexport::Serialize;
+    (@inner $set:ident $this:expr, ($num:expr) ($($processed:tt)*);) => {
+        __enum_to_vec_id!(@create $set $this, ($($processed)*) );
+    };
+}
 
-                match self {$(
-                    &$transaction_set::$name(ref tx) => Serialize::serialize(tx, serializer),
-                )*}
-            }
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __enum_from_id_vec {
+    ($set:ident $this:tt, $($field_type:ident)*) => {
+        __enum_from_id_vec!(@inner $set $this, (0) (); $($field_type)* );
+    };
+    (@create $set:ident ($id:expr, $vec:expr), ( $((($name:ident) => ($num:expr)))* ) ) => {
+        match $id {
+            $(
+                num if num == $num => return <$name as $crate::messages::BinaryForm>::decode(&$vec).map($set::$name),
+            )*
+            num => Err($crate::encoding::Error::Basic(format!("Tag {} not found for enum {}.",num, stringify!($set)).into()))
         }
     };
+    (@inner $set:ident $this:tt, ($num:expr) ($($processed:tt)*);
+        $field_type:ident $($rest:tt)*
+    ) => {
+        __enum_from_id_vec!(
+            @inner $set $this,
+            ($num + 1)
+            ($($processed)* (($field_type) => ($num)));
+            $($rest)*
+        );
+    };
+
+    (@inner $set:ident $this:tt, ($num:expr) ($($processed:tt)*);) => {
+        __enum_from_id_vec!(@create $set $this, ($($processed)*) );
+    };
```

Comment:
No idea what happens in this macro playpen, I hope we check that it works 🤞

Reply:
I don't think that we need tests for this particular macro in isolation. Tests for `transactions!` macro should be enough to verify that this implementation detail works as expected.